x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessAuthenticationService.java

```diff
@@ -0,0 +1,161 @@
+import org.elasticsearch.xpack.core.security.user.SystemUser;
+import org.elasticsearch.xpack.core.security.user.User;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.core.security.authc.RemoteAccessAuthentication.REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY;
+import static org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor.REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY;
+
+public class RemoteAccessAuthenticationService {
+
+    private static final RoleDescriptor CROSS_CLUSTER_SEARCH_ROLE = new RoleDescriptor(
+        "_cross_cluster_search",
+        new String[] { ClusterStateAction.NAME },
+        null,
+        null,
+        null,
+        null,
+        null,
+        null,
+        null
+    );
+    private static final Logger logger = LogManager.getLogger(RemoteAccessAuthenticationService.class);
+
+    private final AuthenticationService authenticationService;
+
+    public RemoteAccessAuthenticationService(AuthenticationService authenticationService) {
+        this.authenticationService = authenticationService;
+    }
+
+    public void authenticate(
+        final String action,
+        final TransportRequest request,
+        final boolean allowAnonymous,
+        final ActionListener<Authentication> listener
+    ) {
+        final Authenticator.Context authcContext = authenticationService.newContext(action, request, allowAnonymous);
+        final ThreadContext threadContext = authcContext.getThreadContext();
+
+        if (threadContext.getHeader(AuthenticationField.AUTHENTICATION_KEY) != null) {
+            withRequestProcessingFailure(
+                authcContext,
+                new IllegalArgumentException("authentication header is not allowed with remote access"),
+                listener
+            );
+            return;
+        }
+
+        final RemoteAccessHeaders remoteAccessHeaders;
+        try {
+            remoteAccessHeaders = RemoteAccessHeaders.readFromContext(threadContext);
+        } catch (Exception ex) {
+            withRequestProcessingFailure(authcContext, ex, listener);
+            return;
+        }
+
+        try (
+            ThreadContext.StoredContext ignored = threadContext.newStoredContext(
+                Collections.emptyList(),
+                // drop remote access authentication headers since we've read their values, and we want to maintain the invariant that
+                // either the remote access authentication header is in the context, or the authentication header, but not both
+                List.of(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY, REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY)
+            )
+        ) {
+            final Supplier<ThreadContext.StoredContext> storedContextSupplier = threadContext.newRestorableContext(true);
```

Comment:
Do we want to preserve response headers? Intuitively, I don't think so since the sub-process `authenticationService.authenticate` should have no other side effect other than provide the authentication result?

Reply:
Good call, I don't think we want to preserve those.